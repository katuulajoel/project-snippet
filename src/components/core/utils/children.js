import React from "react";

export function addPropsToChildren(children, props) {
  return React.Children.map(
    children,
    function (child) {
      return React.cloneElement(child, props);
    }.bind(this)
  );
}

export function addPropsToElement(element, props) {
  return React.cloneElement(element, props);
}
