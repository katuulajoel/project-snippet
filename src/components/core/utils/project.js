import moment from "moment";

export function generateUserIntials(user) {
  return !user.avatar_url ? user.display_name.match(/\b(\w)/g).join("") : null;
}

export const getdateFormated = (date, showTime = false) => {
  var otherDates = moment.utc(date).local().fromNow();
  var calback = function () {
    return showTime ? "[" + otherDates + "], hh:mm A" : "[" + otherDates + "]";
  };
  return moment
    .utc(date)
    .local()
    .calendar(null, {
      sameDay: showTime ? "[Today], hh:mm A" : "[Today]",
      nextDay: calback,
      nextWeek: calback,
      lastDay: calback,
      lastWeek: calback,
      sameElse: showTime ? "Do MMM YYYY, hh:mm A" : "Do MMM YYYY",
    });
};
