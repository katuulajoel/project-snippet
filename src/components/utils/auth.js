// eslint-disable-next-line no-unused-vars
import store from "../../store";

export function getAuth() {
  return store.getState().Auth;
}

export function getUser() {
  return getAuth().user;
}

export function isAdmin() {
  let user = getUser();
  return user.is_admin;
}

export function isPayAdmin() {
  let user = getUser();
  return user.is_pay_admin;
}

export function isDev() {
  return getUser().is_developer;
}

export function isDesigner() {
  return getUser().is_designer;
}

export function isClient() {
  return getUser().is_project_owner;
}

export function isPM() {
  return getUser().is_project_manager && !getUser().is_admin;
}

export function isAdminOrClient() {
  return isAdmin() || isClient();
}

export function isAdminOrPM() {
  return isAdmin() || isPM();
}

export function isPayAdminOrPM() {
  return isPayAdmin() || isPM();
}

export function isAdminOrPMOrClient() {
  return isAdmin() || isPM() || isClient();
}

export function isDevOrClient() {
  return isDev() || isClient();
}

export function isDevOrPM() {
  return isDev() || isPM();
}

export function getMyParticipation(project) {
  let myParticipation = null;
  if (project.participation) {
    (project.participation || []).forEach((item) => {
      if (item.user.id === getUser().id) {
        myParticipation = item;
      }
    });
  }
  return myParticipation;
}

export function isPendingProjectParticipant(project) {
  let isPending = false;
  if (isDev()) {
    let myParticipation = getMyParticipation(project);
    if (myParticipation) {
      return myParticipation.status === "initial";
    }
  }
  return isPending;
}

export function hasProjectAccess(project) {
  let allowedUserIds = [];
  ["user", "owner", "pm"].forEach((key) => {
    if (project[key]) {
      allowedUserIds.push(project[key].id);
    }
  });
  if (project.participation) {
    project.participation.forEach((item) => {
      if (item.status === "accepted" && item.user) {
        allowedUserIds.push(item.user.id);
      }
    });
  }
  return isAdmin() || allowedUserIds.includes(getUser().id);
}

export function isPMAndHasProjectAcess(project) {
  return hasProjectAccess(project) && isPM();
}

export function isAdminOrPMWithProjectAcess(project) {
  return isAdmin() || isPMAndHasProjectAcess(project);
}

export function isAdminOrClientOrPMWithProjectAcess(project) {
  return isAdmin() || isClient() || isPMAndHasProjectAcess(project);
}

export function isProjectClient(project) {
  const userId = getUser().id;
  if (project.owner && project.owner.id === userId) {
    return true;
  }

  if (project.user) {
    return project.user.id === userId && (!project.owner || !isAdmin());
  }
  return false;
}
