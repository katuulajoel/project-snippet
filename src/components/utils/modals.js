import { createConfirmation } from "react-confirm";
import GenericModal from "../GenericModal";

const confirm = createConfirmation(GenericModal);

export function openGenericModal(modalContent, options = {}, modalHeader) {
  return confirm({ modalHeader, modalContent, options: { ...options } });
}

function safeClassNameOptions(className, options) {
  let safeOptions = options || {};
  if (safeOptions.className) {
    safeOptions.className = `${className} ${safeOptions.className}`;
  } else {
    safeOptions.className = className;
  }
  return safeOptions;
}

export function openConfirm({
  message = "",
  title = "",
  canClose = true,
  options = { ok: "Yes", cancel: "Cancel" },
  header = null,
} = {}) {
  return openGenericModal(
    message,
    {
      mustRespond: !canClose,
      title,
      ...safeClassNameOptions("modal-confirm", options),
    },
    header
  );
}

export function openModal({
  body = null,
  title = "",
  canClose = true,
  options = null,
  header = null,
  hideActions = false,
} = {}) {
  return openGenericModal(
    body,
    {
      hideActions,
      mustRespond: !canClose,
      title,
      ...(options || {}),
    },
    header
  );
}
