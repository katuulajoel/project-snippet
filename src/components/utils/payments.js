export const getPaymentsFilters = (filter) => {
  switch (filter) {
    case "paid":
      return { paid: "True" };
    case "pending":
      return { paid: "False", overdue: "False" };
    case "overdue":
      return { overdue: "True", paid: "False" };
    case "archived":
      return { archived: "True" };
    default:
      return {};
  }
};
