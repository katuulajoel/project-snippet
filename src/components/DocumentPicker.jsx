/* -------------------------------------------------------------------------- */
/*                            External Dependencies                           */
/* -------------------------------------------------------------------------- */
import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";

/* -------------------------- Internal dependencies ------------------------- */
import DocumentForm from "./DocumentForm";

import { openModal } from "../utils/modals";
import Icon from "./Icon";
import { DOCUMENT_TYPES_MAP } from "../configs/constants/projectConstants";

const DocumentPicker = ({ onChange, showSelected, documentType }) => {
  const [documents, setdocuments] = useState([]);

  useEffect(() => {
    onChange(documents);
  }, [documents]);

  const onSelectDocument = () => {
    openModal({
      body: <DocumentForm id="document-form" />,
      title: `Add ${`${DOCUMENT_TYPES_MAP[documentType]}${
        documentType == "other" ? "s" : ""
      }`}`,
      options: {
        ok: "Add Document",
        form: {
          type: "submit",
          form: "document-form",
        },
      },
    }).then((document) => {
      setdocuments([...documents, { ...document, type: documentType }]);
    });
  };

  const onRemoveDoc = (idx) => {
    setdocuments([...documents.slice(0, idx), ...documents.slice(idx + 1)]);
  };

  return (
    <div className="document-input">
      {showSelected && documents.length > 0 && (
        <div className="file-list">
          {documents.map((doc, index) => {
            return (
              <div className="file-item" key={index}>
                <i className="tg-ic-file" /> {doc.title}{" "}
                {doc.file ? doc.file.name : doc.url}
                <button className="btn" onClick={() => onRemoveDoc(index)}>
                  <i className="tg-ic-close" />
                </button>
              </div>
            );
          })}
        </div>
      )}

      <a href="#" className="add-btn" onClick={() => onSelectDocument()}>
        <Icon name="round-add" size="sm" />
        &nbsp;&nbsp;Add New
      </a>
    </div>
  );
};

DocumentPicker.defaultProps = {
  showSelected: true,
};

DocumentPicker.propTypes = {
  className: PropTypes.string,
  onChange: PropTypes.func,
  documentType: PropTypes.string,
  showSelected: PropTypes.bool,
};

export default DocumentPicker;
