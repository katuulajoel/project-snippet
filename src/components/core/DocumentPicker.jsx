/* -------------------------------------------------------------------------- */
/*                            External Dependencies                           */
/* -------------------------------------------------------------------------- */
import React from "react";
import PropTypes from "prop-types";
import _ from "lodash";

/* -------------------------- Internal dependencies ------------------------- */
import IconButton from "./IconButton";
import DocumentType from "./DocumentType";
import DocumentForm from "./DocumentForm";

import { openModal } from "./utils/modals";
import { addPropsToChildren } from "./utils/children";
import Button from "./Button";
import Icon from "./Icon";
import { DOCUMENT_TYPES_MAP } from "../../actions/utils/api";

const ModalWrapper = props => {
    return (
        <div>
            {addPropsToChildren(props.children, {
                onChange: props.proceed,
                onSave: props.proceed
            })}
        </div>
    );
};

ModalWrapper.propTypes = {
    children: PropTypes.oneOfType([PropTypes.elementType, PropTypes.object]),
    proceed: PropTypes.func
};

export default class DocumentPicker extends React.Component {
    static defaultProps = {
        variant: "icon",
        showSelected: true,
        instruction: ""
    };

    static propTypes = {
        variant: PropTypes.string,
        className: PropTypes.string,
        onChange: PropTypes.func,
        documentType: PropTypes.string,
        documentTypes: PropTypes.array,
        showSelected: PropTypes.bool,
        size: PropTypes.string,
        instruction: PropTypes.string
    };

    constructor(props) {
        super(props);
        this.state = {
            documents: []
        };
    }

    componentDidUpdate(prevProps, prevState) {
        if (
            this.props.onChange &&
            !_.isEqual(prevState.documents, this.state.documents)
        ) {
            this.props.onChange(this.state.documents);
        }
    }

    onSelectType = () => {
        openModal(
            <ModalWrapper>
                <DocumentType />
            </ModalWrapper>,
            "Type of document",
            true,
            { className: "modal-upload" }
        ).then(type => {
            this.onSelectDocument(type);
        });
    };

    onSelectDocument(type) {
        let self = this;
        openModal(
            <ModalWrapper>
                <DocumentForm
                    id="document-form"
                    type={type}
                    documentType={this.props.documentType}
                    documentTypes={this.props.documentTypes}
                />
            </ModalWrapper>,
            `Add ${
                this.props.documentType
                    ? `${DOCUMENT_TYPES_MAP[this.props.documentType]}${
                          this.props.documentType == "other" ? "s" : ""
                      }` || `${this.props.documentType} document`
                    : " document"
            }`,
            true,
            {
                ok: "Add Document",
                form: {
                    type: "submit",
                    form: "document-form"
                }
            },
            null,
            false
        ).then(document => {
            self.setState({ documents: [...this.state.documents, document] });
        });
    }

    onRemoveDoc(idx) {
        this.setState({
            documents: [
                ...this.state.documents.slice(0, idx),
                ...this.state.documents.slice(idx + 1)
            ]
        });
    }

    renderSwitch(param) {
        switch (param) {
            case "button":
                return (
                    <Button onClick={this.onSelectType}>
                        <Icon name="upload" size={this.props.size} />{" "}
                        {this.props.instruction || "Add documents"}
                    </Button>
                );
            case "new-button":
                return (
                    <a
                        href="#"
                        className="add-btn"
                        onClick={() => this.onSelectDocument("url")}
                    >
                        <Icon name="round-add" size="sm" />
                        &nbsp;&nbsp;Add New
                    </a>
                );
            default:
                return (
                    <>
                        <IconButton
                            name="add"
                            size={this.props.size}
                            onClick={this.onSelectType}
                        />{" "}
                        {this.props.instruction}
                    </>
                );
        }
    }

    render() {
        return (
            <div className="document-input">
                {this.props.showSelected && this.state.documents.length > 0 ? (
                    <div className="file-list">
                        {this.state.documents.map((doc, index) => {
                            return (
                                <div className="file-item" key={index}>
                                    <i className="tg-ic-file" /> {doc.title}{" "}
                                    {doc.file ? doc.file.name : doc.url}
                                    <button
                                        className="btn"
                                        onClick={this.onRemoveDoc.bind(
                                            this,
                                            index
                                        )}
                                    >
                                        <i className="tg-ic-close" />
                                    </button>
                                </div>
                            );
                        })}
                    </div>
                ) : null}

                {this.renderSwitch(this.props.variant)}
            </div>
        );
    }
}
