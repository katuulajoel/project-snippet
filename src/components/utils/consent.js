/* -------------------------------------------------------------------------- */
/*                            External Dependencies                           */
/* -------------------------------------------------------------------------- */
import React from "react";
import moment from "moment";
import * as Cookies from "js-cookie";
import styled from "styled-components";
import PropTypes from "prop-types";

/* -------------------------- Internel Dependencies ------------------------- */
import { openModal } from "./modals";
import CookieSettings from "../CookieSettings";

export const COOKIE_OPTIONS = [
  [
    "essential",
    "Essential Website Cookies",
    "These cookies are strictly necessary to provide you with services available through our website and to use some of its features, such as logging in, filling forms and setting your privacy settings. Because these cookies are strictly necessary to deliver the website, you cannot refuse them without impacting how our website functions.",
    true,
    true,
    [
      [
        "Tunga",
        "https://tunga.io",
        "These cookies are strictly necessary to provide you with services through Tunga",
        [
          "sessionid",
          "csrftoken",
          "taskEditToken",
          "cookieConsent",
          "cookieConsentClosedAt",
        ],
      ],
      [
        "Google Tag Manager",
        "https://www.google.com/analytics/tag-manager/",
        "This cookie is associated with Google Tag Manager which we use to load scripts into our website pages.",
        ["_dc_gtm_UA-70644715-1"],
      ],
    ],
  ],
  [
    "perf_func",
    "Performance and Functionality Cookies",
    "These cookies are used to enhance the performance and functionality of our websites but are non-essential to their use. However, without these cookies, certain functionality may become unavailable.",
    true,
    false,
    [
      [
        "Tunga",
        "https://tunga.io",
        "These cookies are used to enhance the performance and functionality of our website but are not essential to its use.",
        ["chatAutoOpenAt"],
      ],
    ],
  ],
  [
    "analytics_custom",
    "Analytics and Customization Cookies",
    "These cookies collect information that is used either in aggregate form to help us understand how our website is being used or how effective our marketing campaigns are, or to help us customize our website for you in order to enhance your experience.",
    true,
    false,
    [
      [
        "Google Analytics",
        "https://analytics.google.com/analytics/web/",
        "Google Analytics gathers information allowing us to understand interactions with our websites and ultimately refine that experience to better serve you.",
        ["_gat", "_ga", "_gid"],
      ],
      [
        "Hotjar",
        "https://www.hotjar.com/",
        "Hotjar gathers information allowing us to understand interactions with our websites and ultimately refine that experience to better serve you.",
        ["_hj*"],
      ],
      [
        "Optimizely",
        "https://www.optimizely.com/",
        "Optimizely is a testing and experimentation platform that helps us uncover customer insights and create optimal web experiences.",
        ["optimizelyDomainTestCookie", "optimizelyEndUserId"],
      ],
    ],
  ],
  [
    "target_advert",
    "Targeting and Advertising Cookies",
    "These cookies are used to make advertising messages more relevant to you and your interests. They also perform functions like preventing the same ad from continuously reappearing, ensuring that ads are properly displayed, and in some cases selecting advertisements that are based on your interests.",
    true,
    false,
  ],
];

export function setCookieConsent(consentDetails) {
  Cookies.set(
    "cookieConsent",
    `${moment.utc().format()}|${consentDetails.join("|")}`,
    { expires: 365 }
  );
}

export function getCookieConsent() {
  return Cookies.get("cookieConsent");
}

export function removeCookieConsent() {
  Cookies.remove("cookieConsent");
}

export function setCookieConsentCloseAt() {
  Cookies.set("cookieConsentClosedAt", moment.utc().format(), {
    expires: 365,
  });
}

export function getCookieConsentCloseAt() {
  return Cookies.get("cookieConsentClosedAt");
}

export function parseDefaultConsents() {
  let currentConsents = [],
    currentConsentCookie = getCookieConsent();
  if (currentConsentCookie) {
    currentConsentCookie.split("|").forEach((category, idx) => {
      if (idx !== 0 && !moment.utc(category).isValid()) {
        currentConsents.push(category);
      }
    });
  } else {
    COOKIE_OPTIONS.forEach((category) => {
      if (category[3] && !category[4]) {
        currentConsents.push(category[0]);
      }
    });
  }
  return currentConsents;
}

export function openCookieConsentPopUp() {
  openModal(
    <CookieSettings />,
    null,
    true,
    {
      className: "consent",
    },
    <ModalHeader />
  );
}

const ModalHeader = ({ dismiss }) => {
  return (
    <StyledHeader>
      <h3>Cookie Settings</h3>
      <span onClick={dismiss}>Done</span>
    </StyledHeader>
  );
};

ModalHeader.propTypes = {
  dismiss: PropTypes.func,
};

const StyledHeader = styled.div`
  h3 {
    font-weight: 700;
    font-size: 14px;
  }

  span {
    position: absolute;
    line-height: 1.5;
    height: auto;
    padding: 0;
    margin: 0;

    top: 15px;
    right: 15px;
    font-style: normal;
    font-weight: 600;
    font-size: 14px;
    color: #8f9bb3;

    &:hover {
      cursor: pointer;
    }
  }
`;
