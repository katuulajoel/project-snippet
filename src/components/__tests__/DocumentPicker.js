import { mount } from "enzyme/build";
import React from "react";
import renderer from "react-test-renderer";
import DocumentPicker from "../DocumentPicker";
import * as modalActions from "../../utils/modals";
import { act } from "@testing-library/react";
// import { render, screen, fireEvent, act } from "@testing-library/react";

describe("Document picker test", () => {
  it("Snapshot test for Login component", () => {
    const tree = renderer.create(<DocumentPicker />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should select document", () => {
    const onChangeSpy = jest.fn();
    jest.spyOn(modalActions, "openModal").mockReturnValue(Promise.resolve({}));
    act(() => {
      const wrapper = mount(<DocumentPicker onChange={onChangeSpy} />);
      wrapper.find("a").simulate("click");
      console.log(wrapper.debug());
    });

    /* act(() => {
      const { getByText } = render(<DocumentPicker onChange={onChangeSpy} />);
      const link = getByText("Add New");
      fireEvent.click(link);
    });
    screen.debug(); */
  });
});
