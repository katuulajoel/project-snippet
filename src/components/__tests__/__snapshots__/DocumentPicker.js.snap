// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document picker test Snapshot test for Login component 1`] = `
<div
  className="document-input"
>
  <a
    className="add-btn"
    href="#"
    onClick={[Function]}
  >
    <i
      className="tg-ic-round-add tunga-ic-sz-sm "
    />
      Add New
  </a>
</div>
`;
